crates/dojo-core/src/world/world_contract.cairo

```diff
@@ -489,96 +548,136 @@ pub mod world {
             EventEmitter::emit(ref self, WriterUpdated { resource, contract, value: false });
         }
 
-        /// Registers a model in the world. If the model is already registered,
-        /// the implementation will be updated.
+        /// Registers a model in the world.
         ///
         /// # Arguments
+        /// * `namespace` -
+        /// * `name` -
         ///
-        /// * `class_hash` - The class hash of the model to be registered.
-        fn register_model(ref self: ContractState, class_hash: ClassHash) {
+        fn register_model(ref self: ContractState, model_definition: dojo::model::ModelDefinition) {
             let caller = get_caller_address();
 
-            let salt = self.models_salt.read();
-            let (address, name, selector, namespace, namespace_hash) =
-                dojo::model::deploy_and_get_metadata(
-                salt.into(), class_hash
-            )
-                .unwrap_syscall();
-            self.models_salt.write(salt + 1);
+            let namespace_selector = bytearray_hash(@model_definition.namespace);
 
-            if selector.is_zero() {
-                panic_with_byte_array(@errors::invalid_resource_selector(selector));
+            if !self.is_namespace_registered(namespace_selector) {
+                panic_with_byte_array(
+                    @errors::namespace_not_registered(@model_definition.namespace)
+                );
             }
+            self
+                .assert_caller_namespace_write_access(
+                    @model_definition.namespace, namespace_selector
+                );
 
-            if !self.is_namespace_registered(namespace_hash) {
-                panic_with_byte_array(@errors::namespace_not_registered(@namespace));
-            }
+            let model_selector = selector_from_names(
```

Comment:
hum, we have this hash here which may actually be also step intensive when in a contract it's already flatten.
May be worth to save the selector and the namespace hash to avoid re-computing them?

We need to measure this difference between reading a storage felt252 and making a contract call with no arguments (which is what we are doing currently).

Reply:
In fact, I just added the model selector in the `ModelDefinition` struct so the hash computing is now useless. Thanks 🙏